src/main/scala/li/cil/oc/server/component/traits/FluidContainerTransfer.scala

```diff
@@ -0,0 +1,197 @@
+                  FluidContainerUtils.fluidHandlerIn(sink, checkSinkSlot(sink)),
+                  (sourceContainer, sinkContainer) => FluidUtils.transferBetweenFluidHandlers(sourceContainer, ForgeDirection.UNKNOWN, sinkContainer, ForgeDirection.UNKNOWN, count),
+                  (sourceContainer, sinkContainer) => {
+                    val sourceResult = FluidContainerUtils.getContainerResult(sourceContainer)
+                    val sinkResult = FluidContainerUtils.getContainerResult(sinkContainer)
+                    if (
+                      syncResult(source, sourceSide, checkSourceSlot(source), sourceOutput, sourceOutputSide, checkSourceOutputSlot(sourceOutput), sourceResult, simulate = true)
+                        && syncResult(sink, sinkSide, checkSinkSlot(sink), sinkOutput, sinkOutputSide, checkSinkOutputSlot(sinkOutput), sinkResult, simulate = true)
+                    ) {
+                      syncResult(source, sourceSide, checkSourceSlot(source), sourceOutput, sourceOutputSide, checkSourceOutputSlot(sourceOutput), sourceResult)
+                      syncResult(sink, sinkSide, checkSinkSlot(sink), sinkOutput, sinkOutputSide, checkSinkOutputSlot(sinkOutput), sinkResult)
+                      true
+                    } else {
+                      false
+                    }
+                  }
+                )
+              })
+            })
+          })
+        })
+    }
+  }
+
+  private def syncResult(inventory: IInventory, inventorySide: ForgeDirection, inventorySlot: Int, output: IInventory, outputSide: ForgeDirection, outputSlot: Option[Int], result: ItemStack, simulate: Boolean = false) = {
+    val stack = if (simulate) result.copy() else result
+    def decrStackSizeIfInserted(inserted: Boolean): Boolean = {
+      if (inserted && !simulate) {
+        inventory.decrStackSize(inventorySlot, 1)
+      }
+      inserted
+    }
+
+    def replaceOr(f: () => Boolean) = {
+      if (inventorySide == outputSide && outputSlot.getOrElse(inventorySlot) == inventorySlot && inventory.getStackInSlot(inventorySlot).stackSize == 1) {
+        if (!simulate) inventory.setInventorySlotContents(inventorySlot, stack)
+        true
+      }
+      else f()
+    }
+
+    outputSlot match {
+      case None =>
+        replaceOr(
+          () => decrStackSizeIfInserted(InventoryUtils.insertIntoInventory(stack, output, Some(outputSide.getOpposite), simulate = simulate))
+        )
+      case Some(slot) =>
+        replaceOr(
+          () => decrStackSizeIfInserted(InventoryUtils.insertIntoInventorySlot(stack, output, Some(outputSide.getOpposite), slot, simulate = simulate))
+        )
+    }
+  }
+
+  private def withReplayableMove(handlerA: Option[IFluidHandler], handlerB: Option[IFluidHandler], moveFunc: (IFluidHandler, IFluidHandler) => Int, cb: (IFluidHandler, IFluidHandler, IFluidHandler => Unit, IFluidHandler => Unit) => Boolean) = {
```

Comment:
`cb` parameter name probably should be expanded a bit, scala code is not very readable at its best, and Forge fluid API does not help any

Reply:
kk, modified